### PR TITLE
[MI-2267]:Wrote test cases for adding configs through slash commands

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**

### CODE REVIEW

File: server/confluence_config_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	}{
+		"success": {
+			method:     http.MethodPost,
+			statusCode: http.StatusOK,
+			body: `{
+				"type": "dialog_submission",
+				"callback_id": "callbackID",
+				"state": "", 
+				"submission": {
+					"Client ID": "pppppppppp",
```
Comment (ayusht2810) [inline server/confluence_config_test.go]: we can give something like mockClientID

File: server/confluence_config_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+				})
+			},
+		},
+		"wrong api method": {
+			method:     http.MethodGet,
+			statusCode: http.StatusMethodNotAllowed,
+		},
+		"invalid body": {
+			method:     http.MethodPost,
+			statusCode: http.StatusInternalServerError,
```
Comment (ayusht2810) [inline server/confluence_config_test.go]: shouldn't this be a Bad Request?

File: server/config_list_test.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	mockAPI := baseMock()
+	mockAPI.On("LogError", mockAnythingOfTypeBatch("string", 13)...).Return(nil)
+
+	mockAPI.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...).Return(nil)
+
+	p := Plugin{}
+	p.SetAPI(mockAPI)
+
+	mockAPI.On("GetBundlePath").Return("/test/testBundlePath", nil)
```
Comment (raghavaggarwal2308) [inline server/config_list_test.go]:
```suggestion
	mockAPI.On("LogError", mockAnythingOfTypeBatch("string", 13)...)
	mockAPI.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...)
	mockAPI.On("GetBundlePath").Return("/test/testBundlePath", nil)

	p := Plugin{}
	p.SetAPI(mockAPI)
```

File: server/config_list_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		t.Run(name, func(t *testing.T) {
+			defer monkey.UnpatchAll()
+
+			mockAPI.On("GetUser", mock.AnythingOfType("string")).Return(&model.User{Id: "123", Roles: "system_admin"}, nil)
+
+			if tc.patchFuncCalls != nil {
+				tc.patchFuncCalls()
+			}
+
+			request := httptest.NewRequest(tc.method, "/api/v1/autocomplete/configs", nil)
```
Comment (raghavaggarwal2308) [inline server/config_list_test.go]:
```suggestion
			request := httptest.NewRequest(tc.method, "/api/v1/autocomplete/configs", nil)
```
Is there a constant for this path?

File: server/config_list_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+			request.Header.Set(config.HeaderMattermostUserID, "test-user")
+			w := httptest.NewRecorder()
+			p.ServeHTTP(&plugin.Context{}, w, request)
+			assert.Equal(t, tc.statusCode, w.Result().StatusCode)
```
Comment (raghavaggarwal2308) [inline server/config_list_test.go:76]: Can we add a check for the response body as well?

File: server/confluence_config.go [hunk 6]
```diff
@@ -18,11 +18,17 @@ func (p *Plugin) handleConfluenceConfig(w http.ResponseWriter, r *http.Request)
 	pathParams := mux.Vars(r)
 	channelID := pathParams[ParamChannelID]
 	userID := pathParams[ParamUserID]
+	if !model.IsValidId(channelID) || !model.IsValidId(userID) {
+		http.Error(w, "invalid channel or user id", http.StatusBadRequest)
+		p.API.LogError("invalid channel or user id")
```
Comment (raghavaggarwal2308) [inline server/confluence_config.go]:
```suggestion
		http.Error(w, "Invalid channel or user id", http.StatusBadRequest)
		p.API.LogError("Invalid channel or user id")
```

File: server/confluence_config_test.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	mockAPI := baseMock()
+	mockAPI.On("LogError", mockAnythingOfTypeBatch("string", 13)...).Return(nil)
+
+	mockAPI.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...).Return(nil)
+
+	p := Plugin{}
+	p.SetAPI(mockAPI)
+
+	mockAPI.On("GetBundlePath").Return("/test/testBundlePath", nil)
```
Comment (raghavaggarwal2308) [inline server/confluence_config_test.go]:
```suggestion
	mockAPI.On("LogError", mockAnythingOfTypeBatch("string", 13)...).Return(nil)
	mockAPI.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...).Return(nil)
	mockAPI.On("GetBundlePath").Return("/test/testBundlePath", nil)

	p := Plugin{}
	p.SetAPI(mockAPI)
```

File: server/confluence_config_test.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+	p.userStore = getMockUserStoreKV()
+	p.instanceStore = p.getMockInstanceStoreKV(1)
+
```
Comment (raghavaggarwal2308) [inline server/confluence_config_test.go]:
```suggestion
```

File: server/confluence_config_test.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	p.otsStore = p.getMockOTSStoreKV()
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			defer monkey.UnpatchAll()
+
+			mockAPI.On("SendEphemeralPost", mock.AnythingOfType("string"), mock.AnythingOfType("*model.Post")).Once().Return(&model.Post{})
+
```
Comment (raghavaggarwal2308) [inline server/confluence_config_test.go]:
```suggestion
```

File: server/confluence_config_test.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+			mockAPI.On("GetUser", mock.AnythingOfType("string")).Return(&model.User{Id: "123", Roles: "system_admin"}, nil)
+
+			if tc.patchFuncCalls != nil {
+				tc.patchFuncCalls()
+			}
+
+			request := httptest.NewRequest(tc.method, fmt.Sprintf("/api/v1/config/%s/%s", tc.channelID, tc.userID), bytes.NewBufferString(tc.body))
+
```
Comment (raghavaggarwal2308) [inline server/confluence_config_test.go]:
```suggestion
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/config_list_test.go (hunk 4 above)
Comment (Kshitij-Katiyar) [reply server/config_list_test.go]: no have to create for this i think this is not going to be used anywhere else so no need to create the constant

File: server/config_list_test.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+			},
+		},
+		"wrong api method": {
+			method:     http.MethodPost,
+			statusCode: http.StatusMethodNotAllowed,
+		},
+	}
+	mockAPI := baseMock()
+	mockAPI.On("LogError", mockAnythingOfTypeBatch("string", 13)...).Return(nil)
+	mockAPI.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...).Return(nil)
```
Comment (raghavaggarwal2308) [inline server/config_list_test.go]:
```suggestion
	mockAPI.On("LogError", mockAnythingOfTypeBatch("string", 13)...)
	mockAPI.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...)
```
Check everywhere

**Review by raghavaggarwal2308 — APPROVED** [review]

**Review by ayusht2810 — APPROVED** [review]

File: server/command.go [hunk 12]
```diff
@@ -269,11 +270,17 @@ func addConfig(p *Plugin, context *model.CommandArgs, args ...string) *model.Com
 		return &model.CommandResponse{}
 	}
 
+	var defaultServerURL string
```
Comment (manojmalik20) [inline server/command.go]:
```suggestion
	defaultServerURL := ""
```

File: server/config_list_test.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHandleGetConfigList(t *testing.T) {
+	tests := map[string]struct {
+		method         string
+		statusCode     int
+		patchFuncCalls func()
+		resp           []model.AutocompleteListItem
```
Comment (manojmalik20) [inline server/config_list_test.go]: Can't we modify it to `[]*model.AutocompleteListItem`?

File: server/config_list_test.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	}{
+		"success": {
+			method:     http.MethodGet,
+			statusCode: http.StatusOK,
+			patchFuncCalls: func() {
+				monkey.PatchInstanceMethod(reflect.TypeOf(&Plugin{}), "GetConfigKeyList", func(_ *Plugin) ([]string, error) {
+					return []string{
+						"https://test.com",
+					}, nil
```
Comment (manojmalik20) [inline server/config_list_test.go]:
```suggestion
					return []string{"https://test.com"}, nil
```

File: server/config_list_test.go [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+				})
+			},
+			resp: []model.AutocompleteListItem{
+				{
+					Item: "https://test.com",
+				},
+			},
+		},
+		"wrong api method": {
```
Comment (manojmalik20) [inline server/config_list_test.go]:
```suggestion
		"wrong API method": {
```

File: server/config_list_test.go [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+			statusCode: http.StatusMethodNotAllowed,
+		},
+	}
+	mockAPI := baseMock()
+	mockAPI.On("LogError", mockAnythingOfTypeBatch("string", 13)...)
+	mockAPI.On("LogDebug", mockAnythingOfTypeBatch("string", 11)...)
+	mockAPI.On("GetBundlePath").Return("/test/testBundlePath", nil)
+
+	p := Plugin{}
+	p.SetAPI(mockAPI)
```
Comment (manojmalik20) [inline server/config_list_test.go]: You should create a function for this.

File: server/config_list_test.go [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+		t.Run(name, func(t *testing.T) {
+			defer monkey.UnpatchAll()
+
+			mockAPI.On("GetUser", mock.AnythingOfType("string")).Return(&model.User{Id: "123", Roles: "system_admin"}, nil)
+
+			if tc.patchFuncCalls != nil {
+				tc.patchFuncCalls()
+			}
+
+			request := httptest.NewRequest(tc.method, "/api/v1/autocomplete/configs", nil)
```
Comment (manojmalik20) [inline server/config_list_test.go]: Use a constant for the path.

File: server/confluence_config_test.go [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+				"state": "", 
+				"submission": {
+					"Client ID": "mock-ClientID",
+					"Client Secret": "mock-ClientSecret",
+					"Server URL": "https://test.com"
+				},
+				"cancelled": false
+			}`,
+			userID:    "iu73atknztnctef8b8ey9gm6zc",
+			channelID: "tgniw3kmrjd93qns11cboditme",
```
Comment (manojmalik20) [inline server/confluence_config_test.go]: You should create testutil functions for these.

File: server/confluence_config_test.go [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+			patchFuncCalls: func() {
+				monkey.PatchInstanceMethod(reflect.TypeOf(&Plugin{}), "GetConfigKeyList", func(_ *Plugin) ([]string, error) {
+					return []string{
+						"https://test.com",
+					}, nil
+				})
+			},
+		},
+		"wrong api method": {
```
Comment (manojmalik20) [inline server/confluence_config_test.go]:
```suggestion
		"wrong API method": {
```
Fix this everywhere.

File: server/confluence_config_test.go [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+			method:     http.MethodPost,
+			statusCode: http.StatusBadRequest,
+			body:       `{`,
+			userID:     "iu73atknztnctef8b8ey9gm6zc",
+			channelID:  "tgniw3kmrjd93qns11cboditme",
+			patchFuncCalls: func() {
+				monkey.PatchInstanceMethod(reflect.TypeOf(&Plugin{}), "GetConfigKeyList", func(_ *Plugin) ([]string, error) {
+					return []string{
+						"https://test.com",
+					}, nil
```
Comment (manojmalik20) [inline server/confluence_config_test.go]:
```suggestion
					return []string{"https://test.com"}, nil
```
Also change this everywhere.

File: server/confluence_config_test.go [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+	p.userStore = getMockUserStoreKV()
+	p.instanceStore = p.getMockInstanceStoreKV(1)
+	p.otsStore = p.getMockOTSStoreKV()
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			defer monkey.UnpatchAll()
+
+			mockAPI.On("SendEphemeralPost", mock.AnythingOfType("string"), mock.AnythingOfType("*model.Post")).Once().Return(&model.Post{})
+			mockAPI.On("GetUser", mock.AnythingOfType("string")).Return(&model.User{Id: "123", Roles: "system_admin"}, nil)
```
Comment (manojmalik20) [inline server/confluence_config_test.go]: You should also create a testutil function for getting this `model.User`.

File: server/plugin_test.go [hunk 22]
```diff
@@ -95,3 +96,216 @@ func TestExecuteCommand(t *testing.T) {
 		})
 	}
 }
+
+func TestConfigCommand(t *testing.T) {
+	p := Plugin{}
+	mockAPI := &plugintest.API{}
+	p.API = mockAPI
+	for name, val := range map[string]struct {
```
Comment (manojmalik20) [inline server/plugin_test.go]:
```suggestion
	for name, test := range map[string]struct {
```

File: server/plugin_test.go [hunk 23]
```diff
@@ -95,3 +96,216 @@ func TestExecuteCommand(t *testing.T) {
...
+					assert.Equal(t, val.ephemeralMessage, post.Message)
+				}
+			}).Once().Return(&model.Post{})
+
+			roles := "system_user"
+			if val.isAdmin {
+				roles += " system_admin"
+			}
+
+			mockAPI.On("GetUser", mock.AnythingOfType("string")).Return(&model.User{Id: "123", Roles: roles}, nil)
```
Comment (manojmalik20) [inline server/plugin_test.go]: Ensure you use the testutil function here and everywhere after creating the testutil function.

File: server/plugin_test.go [hunk 24]
```diff
@@ -95,3 +96,216 @@ func TestExecuteCommand(t *testing.T) {
...
+
+			res, err := p.ExecuteCommand(&plugin.Context{}, val.commandArgs)
+			assert.Nil(t, err)
+			assert.NotNil(t, res)
+		})
+	}
+}
+
+func TestConfigListCommands(t *testing.T) {
```
Comment (manojmalik20) [inline server/plugin_test.go]:
```suggestion
func TestConfigListCommand(t *testing.T) {
```

File: server/plugin_test.go [hunk 25]
```diff
@@ -95,3 +96,216 @@ func TestExecuteCommand(t *testing.T) {
...
+	p.API = mockAPI
+	p.instanceStore = p.getMockInstanceStoreKV(1)
+	for name, val := range map[string]struct {
+		commandArgs      *model.CommandArgs
+		ephemeralMessage string
+		isAdmin          bool
+		patchAPICalls    func()
+	}{
+		"admin restriction on config list command": {
+			commandArgs:      &model.CommandArgs{Command: "/confluence config ", UserId: userID, ChannelId: channelID},
```
Comment (manojmalik20) [inline server/plugin_test.go]:
```suggestion
			commandArgs:      &model.CommandArgs{Command: "/confluence config list", UserId: userID, ChannelId: channelID},
```

File: server/plugin_test.go [hunk 26]
```diff
@@ -95,3 +96,216 @@ func TestExecuteCommand(t *testing.T) {
...
+			commandArgs:      &model.CommandArgs{Command: "/confluence config \"list\"", UserId: userID, ChannelId: channelID},
+			ephemeralMessage: noSavedConfig,
+			isAdmin:          true,
+			patchAPICalls: func() {
+				monkey.PatchInstanceMethod(reflect.TypeOf(&Plugin{}), "GetConfigKeyList", func(_ *Plugin) ([]string, error) {
+					return []string{}, nil
+				})
+			},
+		},
+		"config list command": {
```
Comment (manojmalik20) [inline server/plugin_test.go]:
```suggestion
		"config list command success": {
```
Fix this everywhere.

File: server/plugin_test.go [hunk 27]
```diff
@@ -95,3 +96,216 @@ func TestExecuteCommand(t *testing.T) {
...
+		isAdmin          bool
+		patchAPICalls    func()
+	}{
+		"admin restriction on config list command": {
+			commandArgs:      &model.CommandArgs{Command: "/confluence config ", UserId: userID, ChannelId: channelID},
+			ephemeralMessage: installOnlySystemAdmin,
+			isAdmin:          false,
+		},
+		"config list command no config saved": {
+			commandArgs:      &model.CommandArgs{Command: "/confluence config \"list\"", UserId: userID, ChannelId: channelID},
```
Comment (manojmalik20) [inline server/plugin_test.go]: Why is "list" is in double quotes in the command?

File: server/plugin_test.go [hunk 28]
```diff
@@ -135,13 +135,13 @@ func TestConfigCommand(t *testing.T) {
 			})
 
 			roles := "system_user"
-			if val.isAdmin {
+			if test.isAdmin {
 				roles += " system_admin"
 			}
 
 			mockAPI.On("GetUser", mock.AnythingOfType("string")).Return(&model.User{Id: "123", Roles: roles}, nil)
```
Comment (manojmalik20) [inline server/plugin_test.go]: You can create a function like `GetMockUser(isAdmin bool)` and call that function here.

File: server/test_utils.go [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+	p.SetAPI(mockAPI)
+
+	return mockAPI, &p
+}
+
+func getValidUserAndChannelID() (string, string) {
+	return validUserID, validChannelID
+}
+
+func getMockUser() *model.User {
```
Comment (manojmalik20) [inline server/test_utils.go]:
```suggestion
func getMockUser(isAdmin bool) *model.User {
```
Then, you can add the roles based on the bool variable.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**